### PR TITLE
[cuda] Nestest Superscale Quantization for CUDA Weights (int4 / int6)

### PR DESCRIPTION
**Pull request by Gasoonjia**
## Quant schema update

Weights are loaded from GGUF and kept at 4-bit (Q4_K) / 6-bit (Q6_K); we further-quantize the per-group metadata (the affine `scale`/`zero`) instead of storing it in bf16. Dequant stays affine: `w = (q - zero) * scale`. The grouping mirrors GGUF's Q4_K/Q6_K layout: `group_size = 32` (int4) / `16` (int6), and a **256-weight super-block** that shares one fp16 metadata step.

- **int4 (Q4_K):** `group_size=32`, super-block=256 (8 groups). `scale` = per-group uint8 code x per-256 fp16 step; `zero` = per-group uint8 code x per-256 fp16 step. Both fp16 steps are packed into one 32-bit word and broadcast to the 8 lanes of a super-block with a single warp shuffle in the decode kernel. **4.625 bpw.**
- **int6 (Q6_K):** symmetric (no zero, exactly like Q6_K), `group_size=16`, super-block=256 (16 groups). `scale` = per-group int8 code x per-256 fp16 step; the per-group int8 codes are hoisted (loaded once per uint4) in the decode kernel. **6.5625 bpw.**

## Why this design

**(a) Follow Q4_K / Q6_K.** Weights come from GGUF, so we mirror llama.cpp's 256-weight super-block granularity (one fp16 step per 256 weights) and its symmetric-Q6_K / asymmetric-Q4_K split, keeping our representation close to the source format.

**(b) int6 keeps its int8 scale code.** int6 is symmetric, so `scale` is its only metadata; dropping it to a 6-bit code collapses accuracy (-9.5 dB) with no zero to compensate, so int6 keeps int8 scale + per-256 fp16 step. A dedicated int6 6-bit-planar-metadata path could revisit this in the future.

## Ablation study

int4 / Q4_K whole-weight SNR vs the exact GGUF Q4_K dequant (same 8-tensor sample throughout; anchors reproduce exactly).

| scale code | scale step | zero code | zero step | SNR (dB) | bpw |
|---|---|---|---|---|---|
| bf16 (no code quant) | - | bf16 | - | 48.04 | - |
| **uint8** | **per-256 fp16** | **uint8** | **per-256 fp16**  (latest / z_pack) | **45.89** | 4.625 |
| uint8 | per-256 fp16 | uint8 | per-256 bf16 | 45.67 | 4.625 |
| uint8 | per-256 fp16 | uint8 | per-row fp16 | 45.17 | 4.5625 |
| uint8 | per-256 fp16 | uint8 | per-row bf16 | 45.15 | 4.5625 |
| uint8 | per-256 bf16 | uint8 | per-row bf16 | 45.10 | 4.5625 |
| uint8 | per-row bf16 | uint8 | per-row bf16  | 44.37 | 4.500 |
| uint8 | per-row fp16 | uint8 | per-row bf16 | 44.36 | 4.500 |
| 6-bit | per-256 fp16 | 8-bit | per-256 fp16 | 46.19* | 4.5625 |
| 6-bit | per-256 fp16 | 6-bit | per-256 fp16 | 38.5 | 4.500 |
| 6-bit | per-row | 8-bit | per-row | 37.85 | 4.441 |

\*6-bit-scale variant: higher raw SNR but rejected on perf (6-bit planar scale reconstruct = +5.2% decode; see below).

**Analysis.** The accuracy lever is **step granularity**, not step dtype or code width: moving the scale step from per-row to per-256 buys +0.78 dB (44.37 -> 45.15), and moving the zero step to per-256 as well buys another +0.74 dB (-> 45.89); step dtype (fp16 vs bf16) is worth <=0.05 dB for a uint8 code. We therefore keep both codes at **uint8** (a 6-bit code loses 6-10 dB, and its planar reconstruct also costs decode perf) and spend the accuracy budget on **per-256 fp16 steps for both scale and zero** — which is exactly GGUF Q4_K's per-super-block-fp16-`d` granularity. This keeps the weight codes byte-aligned and identical to the source 4-bit/6-bit values, holds bpw in a tight band (4.625, +0.125 over the 4.5 baseline), and lands at 45.89 dB (+1.52 dB over the prior 44.37 baseline). The one higher-raw-SNR alternative (6-bit scale + 8-bit zero, 46.19 dB) was rejected because the 6-bit planar scale reconstruct regresses decode +5.2% for +0.3 dB — not worth it.

## e2e results

gemma4-31b Q4_K_M @131072, greedy, cuda_graph, 3-rep median.

### A100, turboquant OFF / bf16 KV

Comparing **ET latest (this PR)** vs **ET prior (pre-metadata-opt baseline)** vs **llama.cpp**.

#### Prefill (tok/s, higher better)

| prompt len | ET latest | ET prior | llama.cpp |
|---:|---:|---:|---:|
| 512    | 1729.7 | 1701.0 | 1266.2 |
| 2048   | 2438.1 | 2494.5 | 1513.7 |
| 8192   | 2273.7 | 2275.6 | 1514.9 |
| 32768  | 1674.6 | 1710.0 | 1269.1 |
| 130048 | 851.6  | 855.9  | 768.8  |

#### Decode (tok/s, higher better)

| prompt len | ET latest | ET prior | llama.cpp |
|---:|---:|---:|---:|
| 512    | 49.93 | 50.5 | 43.63 |
| 2048   | 47.93 | 48.4 | 42.73 |
| 8192   | 47.42 | 48.1 | 42.12 |
| 32768  | 46.79 | 47.0 | 39.74 |
| 130048 | 36.98 | 37.2 | 33.04 |

#### Peak VRAM (GB, lower better)

| prompt len | ET latest | ET prior | llama.cpp |
|---:|---:|---:|---:|
| 512    | 31.56 | 31.38 | 19.38 |
| 2048   | 31.85 | 31.66 | 19.70 |
| 8192   | 31.88 | 31.69 | 20.17 |
| 32768  | 31.88 | 31.69 | 22.04 |
| 130048 | 31.88 | 31.69 | 29.46 |

**A100 takeaways.** vs **ET prior**: decode neutral (-0.45% to -1.41%, mean ~-0.9%), prefill neutral (+-2.3%), VRAM +~0.19 GB (the expected +0.06 bpw of the per-256 zero step) — i.e. +1.52 dB SNR at negligible runtime cost. vs **llama.cpp**: ET decode is +12-18% faster and ET prefill is faster at every length; llama.cpp uses ~12 GB less VRAM at short/mid lengths (packed Q4_K_M weights vs ET's bf16 runtime weights), a gap that narrows to ~2.4 GB at 130048 as the KV cache dominates.

### RTX 5090, turboquant ON / tq4 KV cache

Comparing **ET latest (New, this PR)** vs **ET prior (Orig, pre-metadata-opt baseline)**, decode d=512.

#### Decode (tok/s, higher better) & Peak VRAM (GiB, lower better)

| prompt len | Orig decode | New decode | Decode improvement | Orig VRAM | New VRAM | VRAM improvement |
|---:|---:|---:|:---:|---:|---:|:---:|
| 512   | 57.16 | 60.78 | +6.33% | 25.06 | 23.29 | -7.06% |
| 2048  | 56.39 | 59.79 | +6.03% | 25.06 | 23.29 | -7.06% |
| 8192  | 55.60 | 58.88 | +5.90% | 25.06 | 23.29 | -7.06% |
| 32768 | 55.29 | 58.60 | +5.99% | 25.06 | 23.29 | -7.06% |

**5090 takeaways.** With turboquant tq4 KV, the metadata-opt is a clear win on both axes: decode **+5.9-6.3%** faster and peak VRAM **-7.06%** (-1.77 GiB) at every measured length. (The VRAM drop here is a turboquant-path effect, not the +0.06 bpw weight-metadata cost seen on the A100 bf16-KV config.)

### CODE REVIEW

**Comment by digantdesai** [comment]
@claude review this PR

File: examples/models/gemma4_31b/cuda_source_transformations.py [hunk 1]
```diff
@@ -206,17 +206,21 @@ def _fused_mlp_forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    so a per-output-row concat on dim 0 is exact: the W4A8 dp4a matvec reads each
+    output row's metadata independently, so out[:N_a] reproduces ``a`` and
+    out[N_a:] reproduces ``b`` bit-for-bit (each row keeps its own steps).
     """
     from executorch.backends.cuda.coalesced_int4_tensor import CudaCoalescedInt4Tensor
 
     return CudaCoalescedInt4Tensor(
         torch.cat([a.qdata, b.qdata], dim=0),
         torch.cat([a.scale, b.scale], dim=0),
+        torch.cat([a.scale_step, b.scale_step], dim=0),
```
Comment (digantdesai) [inline examples/models/gemma4_31b/cuda_source_transformations.py:221]: not for this PR but can we share these source transforms across different models?

File: backends/cuda/quantize_op_dispatch/int4_dispatch.py [hunk 2]
```diff
@@ -41,23 +41,25 @@
 # ---------------------------------------------------------------------------
 
 _lib.define(
-    "int4_plain_mm(Tensor self, Tensor qdata, Tensor scale, Tensor zero, int group_size) -> Tensor"
+    "int4_plain_mm(Tensor self, Tensor qdata, Tensor scale, Tensor scale_step, Tensor zero, Tensor zero_step, int group_size) -> Tensor"
```
Comment (digantdesai) [inline backends/cuda/quantize_op_dispatch/int4_dispatch.py]: what about non k-quant users?

File: backends/cuda/quantize_op_dispatch/int4_dispatch.py [hunk 3]
```diff
@@ -88,25 +94,42 @@ def _dequant_matmul(x, qdata, scale, zero, group_size):
...
+        # Scale: uint8 code * per-256 fp16 step (broadcast over the 8 groups in
+        # each super-block).
+        scale_step_g = s_step.to(dtype).repeat_interleave(groups_per_super, dim=1)
+        s = (sc.to(dtype) * scale_step_g).unsqueeze(-1)
+        # Zero: uint8 code * per-256 fp16 step (broadcast over the 8 groups in
+        # each super-block).
+        zero_step_g = z_step.to(dtype).repeat_interleave(groups_per_super, dim=1)
+        z = (ze.to(dtype) * zero_step_g).unsqueeze(-1)
         w_deq = ((data - z) * s).reshape(rows, K)
         return F.linear(x, w_deq)
```
Comment (digantdesai) [inline backends/cuda/quantize_op_dispatch/int4_dispatch.py:117]: Nit: why do we have Linear in `_dq` fn?

File: backends/cuda/runtime/shims/tests/test_aoti_torch_cuda_int4_plain_mm.cpp [hunk 4]
```diff
@@ -122,324 +133,1288 @@ class AOTITorchInt4PlainMMTest : public ::testing::Test {
...
-TEST_F(AOTITorchInt4PlainMMTest, MultiGroupRandom) {
-  int64_t M = 1, K = 32, N = 4, gs = 16;
-
+// SingleSuperBlock: M=1, N=8, K=256, gs=32, ng=8, n_super=1
+TEST_F(AOTITorchInt4PlainMMTest, SingleSuperBlock) {
+  int64_t M = 1, K = 256, N = 8, gs = 32;
+  int64_t ng = K / gs; // 8
+  int64_t n_super = K / 256; // 1
   // clang-format off
   uint8_t qdata_host[] = {
```
Comment (digantdesai) [inline backends/cuda/runtime/shims/tests/test_aoti_torch_cuda_int4_plain_mm.cpp:157]: check in the script to recreate these

File: backends/cuda/coalesced_int4_tensor.py [hunk 5]
```diff
@@ -37,17 +60,72 @@
...
+    the uint8 re-encoding via :meth:`from_int4_tensor`.
     """
 
-    tensor_data_names = ["qdata", "scale", "zero_point"]
+    tensor_data_names = [
+        "qdata",
+        "scale",
+        "scale_step",
+        "zero_point",
+        "zero_step",
```
Comment (digantdesai) [inline backends/cuda/coalesced_int4_tensor.py]: s/zero_step/zero_point_step - Nit

File: backends/cuda/coalesced_int4_tensor.py [hunk 6]
```diff
@@ -56,23 +134,29 @@ def __new__(
         cls,
         qdata: torch.Tensor,
         scale: torch.Tensor,
+        scale_step: torch.Tensor,
         zero_point: torch.Tensor,
+        zero_step: torch.Tensor,
```
Comment (digantdesai) [inline backends/cuda/coalesced_int4_tensor.py]: Is this Tensor only for k-quants?

**Review by digantdesai — COMMENTED** [review]
Looks good except the fact that I am not sure if we want to use the plain kernel for other 4/6b weights or just k-quant with quantized/nested scales.

File: backends/cuda/quantize_op_dispatch/int4_dispatch.py (hunk 3 above)
Comment (Gasoonjia) [reply backends/cuda/quantize_op_dispatch/int4_dispatch.py:117]: That's for memory-efficient exportation. If we dequant the whole block first then do linear, we need to save the whole dequant tensor into VRAM, which takes ltos of memory if tensor is super large (e.g. lm_head). Here we do the dequant + linear in chunk to reduce the peak memory usage.

`_dq` here maybe not a good name; renaming it as `_unit_dq_mm` for better show target.

File: examples/models/gemma4_31b/cuda_source_transformations.py (hunk 1 above)
Comment (Gasoonjia) [reply examples/models/gemma4_31b/cuda_source_transformations.py:221]: yeah we should move this into intx_dispatcher.py/ there're bunch of structure factor we can do for cuda quant pipeline and its on my list.

**Comment by Gasoonjia** [comment]
Thanks @digantdesai  for review.

Tensor now is k-quant only but the pipeline doesn't. We convert the original hdd int4 and int6 tensor into q4_k and q6_k and reuse k-quant pipeline. The precision are the same and only have ~0.7db weight during the conversion.

**Comment by digantdesai** [comment]
> Thanks @digantdesai for review.
> 
> Tensor now is k-quant only but the pipeline doesn't. We convert the original hdd int4 and int6 tensor into q4_k and q6_k and reuse k-quant pipeline. The precision are the same and only have ~0.7db weight during the conversion.

What about performance? Running Q4_0 as nested k-quant kernel must cost some perf, no?

File: backends/cuda/quantize_op_dispatch/int4_dispatch.py (hunk 3 above)
Comment (digantdesai) [reply backends/cuda/quantize_op_dispatch/int4_dispatch.py:117]: ok I remember the chunking. Yeah name is just a nit.

File: examples/models/gemma4_31b/cuda_source_transformations.py (hunk 1 above)
Comment (digantdesai) [reply examples/models/gemma4_31b/cuda_source_transformations.py:221]: add a TODO claude can do the right thing when it sees a todo when adding something else in the same area

**Review by digantdesai — APPROVED** [review]
stamping it to unblock you

**Comment by Gasoonjia** [comment]
> > Thanks @digantdesai for review.
> > Tensor now is k-quant only but the pipeline doesn't. We convert the original hdd int4 and int6 tensor into q4_k and q6_k and reuse k-quant pipeline. The precision are the same and only have ~0.7db weight during the conversion.
> 
> What about performance? Running Q4_0 as nested k-quant kernel must cost some perf, no?

for kernel-base test perf is even better cuz now bpw is lower.
